crypto/src/hash.rs

```diff
@@ -39,11 +39,13 @@ ffi::ffi_item! {
 }
 
 // NOTE: Hash is FFI serialized as an array (a pointer in a function call, by value when part of a struct)
-iroha_ffi::ffi_type! {unsafe impl Transparent for Hash[[u8; Hash::LENGTH]] validated with {Hash::is_lsb_1} }
+iroha_ffi::ffi_type! {
+    unsafe impl Transparent for Hash {
+        type Target = [u8; Hash::LENGTH];
 
-impl iroha_ffi::option::Niche for Hash {
-    // NOTE: Any value that has lsb=0 is a niche value
-    const NICHE_VALUE: Self::ReprC = [0; Hash::LENGTH];
+        validation_fn=unsafe {Hash::is_lsb_1},
+        niche_value = [0; Hash::LENGTH]
+    }
 }
```

Comment:
do you think there could be any issue with treating Hash as transmutable into [u8; Hash::LENGTH] considering it's a repr(C) struct with two fields?

I'm slightly concerned about the effects on the calling convention, though probably that's not an issue

Reply:
Worth investigating, but shouldn't hold this PR back. FFI is not deployed and we have the luxury of not using it in production